aiida_lsmo/utils/cp2k_utils.py

```diff
@@ -1,54 +1,151 @@
+        self._kinds = None
+
+    @property
+    def atoms(self):
+        """Return ASE atoms instance"""
+        if not self._atoms:
+            self._atoms = self.structure.get_ase()
+        return self._atoms
+
+    @property
+    def initial_magnetizations(self):
+        """ Compute the initial magnetization.
+
+        By default, the initial magnetization is determined based on the element.
+        If oxidation states are provided, the guess may be improved based on those.
+        """
+        from aiida_lsmo.workchains.cp2k_multistage_protocols import INITIAL_MAGNETIZATION, get_default_magnetization
+        if not self._initial_magnetizations:
+            all_atoms = self.atoms.get_chemical_symbols()
+
+            # Use magnetization table
+            initial_magnetizations = [get_default_magnetization(element) for element in all_atoms]
+
+            if self.oxidation_states is None:
+                return initial_magnetizations
+
+            # Update selected choices, if oxidation states provided
+            for index, element, oxidation_state in zip(self.oxidation_states['metal_indices'],
+                                                       self.oxidation_states['metal_symbols'],
+                                                       self.oxidation_states['prediction']):
+                initial_magnetizations[index] = INITIAL_MAGNETIZATION[element]['magnetization'][oxidation_state]
+
+            self._initial_magnetizations = initial_magnetizations
+
+        return self._initial_magnetizations
+
+    def get_multiplicity_section(self):
+        """ Compute the total multiplicity of the structure by summing the atomic magnetizations.
+
+            multiplicity = 1 + sum_i ( natoms_i * magnetization_i ), for each atom_type i
+                         = 1 + sum_i magnetization_j, for each atomic site j
+
+        :returns: dict (for cp2k input)
+        """
+        multiplicity = 1 + sum(self.initial_magnetizations)
+        multiplicity = int(round(multiplicity))
+
+        multiplicity_dict = {'FORCE_EVAL': {'DFT': {'MULTIPLICITY': multiplicity}}}
+        if multiplicity != 1:
+            multiplicity_dict['FORCE_EVAL']['DFT']['UKS'] = True
+        return multiplicity_dict
+
+    def get_kinds_section(self):
+        """ Write the &KIND sections given the structure and the settings_dict"""
+        from aiida_lsmo.workchains.cp2k_multistage_protocols import get_default_magnetization
+
+        if not self._kinds:
+            kinds = []
+            all_atoms = set(self.atoms.get_chemical_symbols())
+            for atom in sorted(all_atoms):
+                kinds.append({
+                    '_': atom,
+                    'BASIS_SET': self.protocol['basis_set'][atom],
+                    'POTENTIAL': self.protocol['pseudopotential'][atom],
+                    'MAGNETIZATION': get_default_magnetization(atom),
+                })
+        else:
+            kinds = [{
+                '_': k['element'],
```

Comment:
This is not correctly printing the kind. 
From some testing for Cu(I)/Cu(II)-HKUST-1 I get:
```
      &KIND Cu
         BASIS_SET DZVP-MOLOPT-SR-GTH-q11
         MAGNETIZATION 0
         POTENTIAL GTH-PBE-q11
      &END KIND
      &KIND Cu
         BASIS_SET DZVP-MOLOPT-SR-GTH-q11
         MAGNETIZATION 1
         POTENTIAL GTH-PBE-q11
      &END KIND
```
Just as a sinde note: CP2K should strip the element, from e.g., CuX, if I remember well.
But to be sure one can add the field `ELEMENT`.
https://manual.cp2k.org/trunk/CP2K_INPUT/FORCE_EVAL/SUBSYS/KIND.html#list_ELEMENT